Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp

```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+    }
+
+    if (m_unifiedOriginStorageLevel <= UnifiedOriginStorageLevel::Basic) {
+        auto cacheStoragePath = CacheStorageManager::cacheStorageOriginDirectory(m_customCacheStoragePath, origin);
+        auto cacheStorageModificationTime = valueOrDefault(FileSystem::fileModificationTime(cacheStoragePath));
+        lastModificationTime = cacheStorageModificationTime > lastModificationTime ? cacheStorageModificationTime : lastModificationTime;
+    }
+
+    if (m_unifiedOriginStorageLevel <= UnifiedOriginStorageLevel::Standard) {
+        auto originFile = manager.path();
+        auto originFileModificationTime = valueOrDefault(FileSystem::fileModificationTime(originFile));
+        lastModificationTime = originFileModificationTime > lastModificationTime ? originFileModificationTime : lastModificationTime;
+    }
+
+    return lastModificationTime;
+}
+
+void NetworkStorageManager::didGetDomainsWithLastAccessedTimeForEviction(std::optional<HashMap<WebCore::RegistrableDomain, WallTime>> domainsWithLastAccessedTime)
+{
+    assertIsCurrent(workQueue());
+
+    HashMap<WebCore::SecurityOriginData, AccessRecord> originRecords;
+    uint64_t totalUsage = 0;
+    for (auto& origin : getAllOrigins()) {
+        auto usage = originStorageManager(origin, ShouldWriteOriginFile::No).quotaManager().usage();
+        totalUsage += usage;
+        WallTime accessTime;
+        if (domainsWithLastAccessedTime)
+            accessTime = domainsWithLastAccessedTime->get(WebCore::RegistrableDomain { origin.topOrigin });
+        else
+            accessTime = lastModificationTimeForOrigin(origin, originStorageManager(origin, ShouldWriteOriginFile::No));
+
+        auto& record = originRecords.ensure(origin.topOrigin, [&] {
+            return AccessRecord { };
+        }).iterator->value;
+        record.usage += usage;
+        if (record.lastAccessTime < accessTime)
+            record.lastAccessTime = accessTime;
+
+        record.clientOrigins.append(origin.clientOrigin);
+        removeOriginStorageManagerIfPossible(origin);
+    }
+
+    m_totalUsage = totalUsage;
+    performEviction(WTFMove(originRecords));
+}
+
+void NetworkStorageManager::performEviction(HashMap<WebCore::SecurityOriginData, AccessRecord>&& originRecords)
+{
+    assertIsCurrent(workQueue());
+
+    m_isEvictionScheduled = false;
+    if (!m_totalQuota || !m_totalUsage || *m_totalUsage <= *m_totalQuota)
+        return;
+
+    Vector<WebCore::SecurityOriginData> sortedOrigins;
+    for (auto& origin : originRecords.keys())
+        sortedOrigins.append(origin);
+
+    std::sort(sortedOrigins.begin(), sortedOrigins.end(), [&](const auto& a, const auto& b) {
+        return originRecords.get(a).lastAccessTime > originRecords.get(b).lastAccessTime;
+    });
+
+    uint64_t deletedOriginCount = 0;
+    while (*m_totalUsage > *m_totalQuota) {
+        ASSERT(!sortedOrigins.isEmpty());
+        auto topOrigin = sortedOrigins.takeLast();
```

Comment:
What guarantees sortedOrigin is not empty?
I would tend to iterate on sortedOrigins and exit early if total usage is less than quota.

Reply:
Because m_totalUsage is calculated by summing up usage of sortedOrigins. Sure.